beacon_node/eth1_chain/src/web3_fetcher.rs

```diff
@@ -0,0 +1,245 @@
+use bls::{PublicKeyBytes, SignatureBytes};
+use ethabi::{decode, ParamType, Token};
+use parking_lot::RwLock;
+use std::collections::BTreeMap;
+use std::marker::Send;
+use std::sync::Arc;
+use types::DepositData;
+use web3::contract::{Contract, Options};
+use web3::futures::{Future, Stream};
+use web3::transports::WebSocket;
+use web3::types::FilterBuilder;
+use web3::types::*;
+use web3::Web3;
+
+use crate::types::{ContractConfig, Eth1DataFetcher};
+
+/// Wrapper around web3 api.
+#[derive(Clone, Debug)]
+pub struct Web3DataFetcher {
+    event_loop: Arc<web3::transports::EventLoopHandle>,
+    /// Websocket transport object. Needed for logs subscription.
+    web3: Arc<web3::api::Web3<web3::transports::ws::WebSocket>>,
+    /// Deposit contract config.
+    contract: ContractConfig,
+}
+
+impl Web3DataFetcher {
+    /// Create a new Web3 object.
+    pub fn new(endpoint: &str, deposit_contract: ContractConfig) -> Web3DataFetcher {
+        let (event_loop, transport) = WebSocket::new(endpoint).unwrap();
+        let web3 = Web3::new(transport);
+        Web3DataFetcher {
+            event_loop: Arc::new(event_loop),
+            web3: Arc::new(web3),
+            contract: deposit_contract,
+        }
+    }
+
+    /// Return filter for subscribing to `DepositEvent` event.
+    fn get_deposit_logs_filter(&self) -> Filter {
+        /// Keccak256 hash of "DepositEvent" in bytes for passing to log filter.
+        const DEPOSIT_CONTRACT_HASH: &str =
+            "649bbc62d0e31342afea4e5cd82d4049e7e1ee912fc0889aa790803be39038c5";
+        let filter = FilterBuilder::default()
+            .address(vec![self.contract.address])
+            .topics(
+                Some(vec![DEPOSIT_CONTRACT_HASH.parse().unwrap()]),
+                None,
+                None,
+                None,
+            )
+            .build();
+        filter
+    }
+}
+
+impl Eth1DataFetcher for Web3DataFetcher {
+    /// Get block_number of current block.
+    fn get_current_block_number(&self) -> Option<U256> {
+        let block_future = self.web3.eth().block_number();
+        let block = block_future.wait().ok()?;
+        Some(block)
+    }
+
+    /// Get block hash at given height.
+    fn get_block_hash_by_height(&self, height: u64) -> Option<H256> {
+        let block_future = self
+            .web3
+            .eth()
+            .block(BlockId::Number(BlockNumber::Number(height)));
+        let block = block_future.wait();
+        block.ok().and_then(|x| x).and_then(|b| b.hash)
+    }
+
+    /// Get `deposit_count` from deposit contract at given eth1 block number.
+    fn get_deposit_count(&self, block_number: Option<BlockNumber>) -> Option<u64> {
+        let contract =
+            Contract::from_json(self.web3.eth(), self.contract.address, &self.contract.abi).ok()?;
```

Comment:
Perhaps we can call this once in the constructor and store the `Contract`?

I'm guessing this involves parsing JSON so it's a (somewhat) non-trivial operation.

Reply:
Yup makes sense!